docs/_tutorials/read-only-json-api.md

```diff
@@ -0,0 +1,307 @@
+
+After building the site, you'd get the following:
+
+```json
+{"title":"The First Post","categories":["update"],"tags":[],"content":"<p>1</p>\n","collection":"posts","date":"2020-12-20 00:00:00 +0100","custom-property":"My custom property"}
+``` 
+
+## Generate Listings
+
+The last step is to generate categories and an index of all posts.
+
+### Define a Page
+
+In `api_generator.rb`, add the following class as a part of `Jekyll` module:
+
+```ruby
+class ListingPage < Page
+  def initialize(site, base, dir, data)
+    @site = site
+    @base = base
+    @dir  = dir
+    @name = 'index.json'
+
+    self.process(@name)
+    self.read_yaml(File.join(base, '_layouts'), 'category.html')
+    self.content = JSON.generate(data)
+  end
+end
+```
+
+* [`Page`](https://www.rubydoc.info/github/jekyll/jekyll/Jekyll/Page) means that `ListingPage` represents a generated page.
+* Override `initialize` method. According to the [documentation](https://www.rubydoc.info/github/jekyll/jekyll/Jekyll%2FPage:initialize):
+  
+> `site` - The Site object. 
+> 
+> `base` - The String path to the source.
+>  
+> `dir` - The String path between the source and the file. 
+> 
+> `name` - The String filename of the file.
+
+* Assign some [instance variables](https://www.ruby-lang.org/en/documentation/faq/8/) that Jekyll will utilize to generate the page.
+* Process the page filename. It will extract the extension and base name.
+* `read_yaml` is only needed to define some instance variable for us. It's just simpler than doing it by hand. You can use any other layout as rendered content will be overwritten.
+* Overwrite content with JSON generated from a [hash](https://ruby-doc.org/core-2.7.2/Hash.html) provided in the `data` argument.
+
+### Generate Pages
+
+The last class you'll write is a [generator]({% link _docs/plugins/generators.md %}) that creates listings for you. It's the last piece of code in this tutorial. It's rather lengthy, but really simple in principle:
+
+```ruby
+class ApiGenerator < Generator
+  safe true
+  priority :normal
+
+  def generate(site)
+    categories = {}
+    posts = {}
+    
+    site.categories.each_key do |category|
+      categories[category] = {}
+      site.categories[category].each_entry do |post|
+        if post.data['draft']
+          continue
+        end
+        inserted_post = post.data.clone
+        title = inserted_post['title']
+        inserted_post.delete('draft')
+        inserted_post.delete('ext')
+        inserted_post.delete('title')
+        inserted_post['url'] = post.url
+        categories[category][title] = inserted_post
+        posts[title] = inserted_post
+      end
+    end
+
+    category_dir = site.config['category_dir'] || 'categories'
+    site.categories.each_key do |category|
+      site.pages << ListingPage.new(site, site.source, File.join(category_dir, category), categories[category])
```

Comment:
The code cuts off around `File.join`.

https://deploy-preview-8521--jekyllrb.netlify.app/tutorials/read-only-json-api/#generate-pages

```suggestion
      site.pages << ListingPage.new(site, site.source,
                                    File.join(category_dir, category), 
                                    categories[category])
```

Or

```ruby
foo.new(
  bar
  baz
  buzz
)
```